versioned_hdf5/slicetools.pyx

```diff
@@ -234,6 +238,125 @@ cpdef build_data_dict(dcpl, raw_data_name: str):
     return data_dict
 
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.infer_types(True)
+cpdef tuple[np.ndarray, np.ndarray] build_slab_indices_and_offsets(
+    dcpl,
+    shape: tuple[int, ...],
+    chunk_size: tuple[int, ...],
+):
+    """Build the slab_indices and slab_offsets arrays for a StagedChangesArray that
+    accesses a versioned virtual dataset.
+
+    :param dcpl: the dataset creation property list of the versioned dataset
+    :return: tuple of (slab_indices, slab_offsets) arrays, each with the same shape as
+    the chunks in the virtual dataset.
+    """
+    cdef hsize_t[NPY_MAXDIMS] chunk_size_arr
+    cdef hsize_t[NPY_MAXDIMS] nchunks_arr
+    cdef hsize_t[NPY_MAXDIMS] h5_start
+    cdef hsize_t[NPY_MAXDIMS] scratch  # Dumping ground for unused outputs
+
+    nchunks = []
+    ndim = len(shape)
+    for j in range(ndim):
+        chunk_size_j: hsize_t = chunk_size[j]
+        assert chunk_size_j > 0
+        chunk_size_arr[j] = chunk_size[j]
+        nchunks_j = ceil_a_over_b(shape[j], chunk_size_j)
+        nchunks_arr[j] = nchunks_j
+        nchunks.append(nchunks_j)
+
+    # Default for missing chunks is to point to slabs[0],
+    # which is a single chunk covered in fill_value
+    cdef np.ndarray slab_indices = np.zeros(nchunks, dtype=np_hsize_t)
+    cdef np.ndarray slab_offsets = np.zeros(nchunks, dtype=np_hsize_t)
+    indices_strides = slab_indices.strides
+    slab_indices_data = slab_indices.data
+    slab_offsets_data = slab_offsets.data
+
+    with phil:
+        # Newly created sparse datasets completely full of fill_value are not virtual.
+        # FIXME Error getting virtual sources on empty virtual dataset
+        # https://github.com/h5py/h5py/issues/1660
+        if dcpl.get_layout() != h5d.VIRTUAL or 0 in shape:
+            return slab_indices, slab_offsets
+
+        dcpl_id: hid_t = dcpl.id
+        virtual_count: size_t = dcpl.get_virtual_count()
+
+        with nogil:
+            # Iterate over all chunks except full ones
+            for chunk_i in range(virtual_count):
+                # Calculate offset of the chunk, IN BYTES, relative to the start of the
+                # flattened buffer of slab_indices and slab_offsets
+                vspace_id = H5Pget_virtual_vspace(dcpl_id, chunk_i)
+                if vspace_id == H5I_INVALID_HID:
+                    raise HDF5Error()
+                try:
+                    vspace_type = H5Sget_select_type(vspace_id)
+                    if vspace_type != H5S_sel_type.H5S_SEL_HYPERSLABS:
+                        raise ValueError(f"Unexpected {vspace_type=}")
```

Comment:
`_spaceid_to_slice` handles this edge case. I cannot think how to possibly generate it short of manually crafting a virtual dataset that is not made by versioned_hdf5. @asmeurer any ideas?

Reply:
Sorry, I need more context to understand the question you are asking here (it's been a while since I've worked on this codebase).

Reply:
4.5 years ago  ~~in a galaxy far away~~, when you originally implemented this function in https://github.com/deshaw/versioned-hdf5/commit/b577ad15c1860f0d3ca00a5fa5581769357a33b7, you covered the edge case of either virtual or source dataspaces being H5S_SEL_ALL instead of H5S_SEL_HYPERSLABS. Shortly afterwards, in https://github.com/deshaw/versioned-hdf5/commit/f3a4e3bd0cd26063c7af36dcb7c31ec37697b20c you added support for H5S_SEL_NONE.

These two use cases never occur in the test suite, and I can't think of any way to reproduce them, short of manually crafting a virtual dataset and then expecting versioned_hdf5 to treat it like a versioned dataset - which IMHO should not be something we care about?

Reply:
My guess is I was just being defensive. I do know I tried using hyperslabs at some point to improve performance, but it turned out that they were slower in the underlying HDF5 library than what we are doing now.